eclipse-che/che-server
# chore: Apply authentication status in the callback url

Signed-off-by: Igor Vinokur <ivinokur@redhat.com>



### What does this PR do?
When processing an authentication callback request set additional error query patameter to the callback url.
How it works:
1. User creates a factory form dashboard.
2. Dasboard requests che-server factory API with no error param in the request url. See https://github.com/eclipse-che/che-dashboard/pull/599
3. Che-server create factory API parses the url for the error param. No error query param means `skipAuthentication=false`. Authentication in progress.
https://github.com/eclipse-che/che-server/blob/74eb0a333d8a4c062fcc99253d4dc3f926c866d2/wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java#L104-L106
4. GitHub shows the authentication page. If User rejects the authentication, authentication status is set to `access_denied`.
5. The error status is added to the redirect url as a query param.
https://github.com/eclipse-che/che-server/blob/74eb0a333d8a4c062fcc99253d4dc3f926c866d2/wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPI.java#L85-L93
6. Dasboard is loaded from the redirect url. Dashboard requests the che-server to create factory by an api request with the error status query param from the redirect url.
https://github.com/eclipse-che/che-dashboard/blob/e2849d9d212c5055885b164ff05422005efc1e7d/packages/dashboard-frontend/src/containers/Loader/Factory/Steps/FetchDevfile/index.tsx#L208-L213
7. Che-server create factory API parses the url for the status param. If `skipAuthentication=true` the authentication flow is skiped and factory creation progress goes further. If `skipAuthentication=false` factory is created in a regular way.
https://github.com/eclipse-che/che-server/blob/74eb0a333d8a4c062fcc99253d4dc3f926c866d2/wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java#L70-L81
### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/21436

### How to test this PR?

1. Deploy Eclipse che with the PR image and specific operator image: `chectl server:deploy -p=openshift  -i=quay.io/ivinokur/che-server:next`
2. Apply the related dashboard pr: https://github.com/eclipse-che/che-dashboard/pull/599
3. Setup [github oauth configuration ](https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-github)
4. create a factory from public GH repo
5. Reject the GitHub Authentication

See: Factory is created without the authentication request.
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
